File: frontend/src/lib/distribution.ts
/** Naming the shape of a column, and saying it.
 *
 *  `src/clickhouse/distribution.rs` measures: it counts rows into buckets — one
 *  per value where there are few, equal-width bins where the column is
 *  continuous, the most common values plus an honest remainder where it is
 *  neither — and has no opinion about the counts. This decides what shape they
 *  make and words it.
 *
 *  The vocabulary is deliberately small. Six shapes, each of which changes what
 *  somebody would do next; a taxonomy with twenty entries would be a quiz rather
 *  than an answer. Everything that does not clearly fit one of them is described
 *  by its own numbers instead of being forced into the nearest name — "spread
 *  across 11 of 16 buckets" says less than "two clusters" and is never wrong.
 *
 *  Two facts about the measurement shape the wording, and both come from the
 *  measuring half's own findings.
 *
 *  - **A tally and a histogram are not the same claim.** Fourteen values that
 *    happen to be powers of two are, as *frequencies*, perfectly even — the long
 *    tail is in the values, not in how often they occur. Binned it would read as
 *    a tail; tallied it reads as fourteen equally common values, which is what
 *    is true. So the sentence always says which of the three was counted.
 *  - **An empty bucket is a fact.** The measuring half fills them in on purpose,
 *    so a gap between two peaks is real and can be relied on here. A shape rule
 *    that ignored empties would call a bimodal column "spread across 16
 *    buckets".
 */

export type Mode = 'tally' | 'bins' | 'top'

export interface Bucket {
  label: string
  rows: number
  from: number | null
  to: number | null
}

export interface Distribution {
  available: boolean
  reason: string | null
  column: string
  type: string
  mode: Mode
  rows: number
  nulls: number
  distinct: number
  buckets: Bucket[]
  tail_rows: number
  tail_values: number
}

/** The named shapes. `mixed` is the honest fallback, not a failure. */
export type Shape =
  | 'empty'
  | 'key'
  | 'single'
  | 'dominant'
  | 'even'
  | 'clustered'
  | 'tail'
  | 'mixed'

/** At or above this share of rows being distinct, the column is an identifier
 *  and has no distribution to show. Measured: `analytics.events.payload` holds
 *  482,212 distinct values in 482,212 rows, and its twelve most common are one
 *  row each — twelve bars of height 1, which is a chart that says nothing. */
const KEYLIKE = 0.95

/** How many times its fair share the listed values have to hold before a
 *  `top`-mode column counts as concentrated.
 *
 *  The rule this replaced compared the listed share against a flat threshold and
 *  got `device_id` wrong: 400 values spread evenly, of which the twelve most
 *  common are 3.0% of the rows — which is exactly 12/400, and even. A power law
 *  over the same 400 would put most of the table in those twelve. Only the ratio
 *  between what they hold and what they would hold if the column were flat can
 *  tell the two apart, because the twelve bars look identical either way. */
const UNFAIR = 3

/** One value holding this share of the rows makes the column that value, for
 *  every practical purpose — a `status` that is 'ok' 99% of the time is a
 *  constant with exceptions, and reads as one. */
const NEARLY_ALL = 0.95

/** And this much makes it the answer without making it the only answer. Half,
 *  because half is already the thing somebody would say about the column — the
 *  first draft put it at 0.6 and reported a value holding exactly 50% of the
 *  table as an unremarkable spread. */
const DOMINANT = 0.5

/** How far the fullest and the emptiest occupied buckets may differ before the
 *  column stops being evenly spread. Two, because a uniform column measured over
 *  a finite table wobbles: `uniform` over 200,000 rows gave 12,400 to 12,600
 *  across sixteen bins, a ratio of 1.02, and a threshold near 1 would call that
 *  uneven. */
const EVEN = 2

/** A peak is a bucket holding at least this share. Below it a bump is noise. */
const PEAK = 0.15

/** Buckets between two peaks that must be near-empty for them to be two clusters
 *  rather than one broad one. */
const VALLEY = 0.02

export function shapeOf(d: Distribution): Shape {
  if (d.rows === 0 || d.buckets.length === 0) return 'empty'
  const counts = d.buckets.map((b) => b.rows)
  const total = d.rows
  const top = Math.max(...counts)

  /* Before every other rule: a column whose values are all different is an
     identifier, and an identifier has no distribution. Every rule below would
     have something to say about it and all of them would be beside the point. */
  if (d.distinct / total >= KEYLIKE) return 'key'

  /* The order is the design, and it was wrong twice before it was right. Each
     rule below is checked before the ones that would also be true of the same
     column, so the column gets its most informative name rather than its first
     matching one. */

  if (top / total >= NEARLY_ALL) return 'single'

  /* Before `dominant`, because a symmetric pair of peaks holds exactly half in
     each — and "`v0` is 50.0% of the rows" is a true sentence that hides the
     fact somebody actually needs, which is that there is nothing in between. */
  const peaks = peakIndexes(counts, total)
  if (peaks.length >= 2 && separated(counts, peaks, total)) return 'clustered'

  /* A `top` reading only ever sees the head, so the question it can answer is
     not "what do these twelve look like" but "do they hold more than their
     share". Evenly spread over four hundred values, or a power law over the
     same four hundred, draw the same twelve bars. */
  if (d.mode === 'top' && d.tail_values > 0) {
    const listed = counts.reduce((a, b) => a + b, 0)
    const fair = d.buckets.length / d.distinct
    if (fair > 0 && listed / total / fair >= UNFAIR) return 'tail'
    return 'even'
  }

  /* Also before `dominant`. A duration whose first bucket is 54% is dominant and
     falling, and the slope is the better sentence — "piled at the low end" tells
     somebody the shape, "bucket 0 holds 54%" tells them one bar. */
  if (d.mode === 'bins' && descending(counts)) return 'tail'

  if (top / total >= DOMINANT) return 'dominant'

  const occupied = counts.filter((n) => n > 0)
  if (
    occupied.length === counts.length &&
    occupied.length >= 3 &&
    top / Math.min(...occupied) <= EVEN
  ) {
    // Only where there are no empties: sixteen buckets of which two are full and
    // even with each other are two clusters, not an even spread.
    return 'even'
  }

  return 'mixed'
}

function peakIndexes(counts: readonly number[], total: number): number[] {
  return counts.map((n, i) => (n / total >= PEAK ? i : -1)).filter((i) => i >= 0)
}

/** Whether the peaks form exactly two groups with a real valley between them.
 *
 *  Two rules, and the first draft had neither. Adjacent peaks are **one** group —
 *  the three buckets at the top of a broad hump are one cluster, not three — and
 *  a valley has to be at least two buckets wide, because a single empty bucket
 *  in a ragged column is noise. Without both, a shape of `3000 0 5000 4000 0
 *  6000 3500 0 0 4000` was reported as two clusters, which is a sentence about a
 *  column that does not exist.
 *
 *  Exactly two, and more than two falls back to `mixed` with its own numbers.
 *  Three clusters is a real thing and a rarer one, and naming it would be a
 *  third rule earning its keep on a shape nobody has yet brought. */
function separated(counts: readonly number[], peaks: readonly number[], total: number): boolean {
  const groups: number[][] = []
  for (const i of peaks) {
    const last = groups[groups.length - 1]
    const previous = last ? last[last.length - 1] : undefined
    if (previous !== undefined && i - previous === 1) last!.push(i)
    else groups.push([i])
  }
  if (groups.length !== 2) return false
  const left = groups[0]!
  const right = groups[1]!
  const between = counts.slice(left[left.length - 1]! + 1, right[0]!)
  return between.length >= 2 && between.every((n) => n / total <= VALLEY)
}

/** Monotone non-increasing, allowing the wobble a real column has. */
function descending(counts: readonly number[]): boolean {
  const first = counts[0] ?? 0
  const last = counts[counts.length - 1] ?? 0
  if (first <= last * 2) return false
  let falls = 0
  for (let i = 1; i < counts.length; i++) if (counts[i]! <= counts[i - 1]!) falls++
  return falls >= counts.length - 2
}

/** What was counted, in one clause — because a bar means three different things
 *  across the three modes and a chart that does not say which is a chart that
 *  can be read three ways. */
export function counted(d: Distribution): string {
  switch (d.mode) {
    case 'tally':
      return `${exact(d.distinct)} ${d.distinct === 1 ? 'value' : 'values'}, each counted`
    case 'bins':
      return `${d.buckets.length} equal buckets across the range`
    case 'top':
      return `the ${d.buckets.length} most common of ${exact(d.distinct)} values`
  }
}

/** The shape, as a sentence somebody would say. */
export function says(d: Distribution): string {
  const total = d.rows
  if (total === 0) {
    return d.nulls > 0
      ? `Every one of the ${exact(d.nulls)} rows is null, so there is no distribution to read.`
      : 'There are no rows to read a distribution from.'
  }
  const top = d.buckets.reduce((a, b) => (b.rows > a.rows ? b : a), d.buckets[0]!)
  const share = (n: number) => `${((n / total) * 100).toFixed(1)}%`

  switch (shapeOf(d)) {
    case 'single':
      return `Nearly every row is ${value(top.label, d)} — ${share(top.rows)} of them.`
    case 'dominant':
      return `${value(top.label, d)} is ${share(top.rows)} of the rows; the rest are spread across ${
        occupiedCount(d) - 1
      } other ${occupiedCount(d) - 1 === 1 ? 'bucket' : 'buckets'}.`
    case 'key':
      return `Every row has its own value — ${exact(d.distinct)} of them in ${exact(
        total,
      )} rows. This is an identifier, not a distribution.`
    case 'even':
      if (d.mode === 'top') {
        return `Evenly spread: the ${d.buckets.length} most common of ${exact(
          d.distinct,
        )} values hold ${share(listedRows(d))} of the rows, which is about their share of the values.`
      }
      return d.mode === 'tally'
        ? `${exact(d.distinct)} values, all about equally common.`
        : `Evenly spread across the range — every bucket holds between ${share(
            Math.min(...d.buckets.map((b) => b.rows)),
          )} and ${share(top.rows)}.`
    case 'clustered':
      return `Two clusters, with nothing between them: ${peakLabels(d).join(' and ')}.`
    case 'tail':
      return d.mode === 'top'
        ? `A long tail: the ${d.buckets.length} most common values are only ${share(
            d.buckets.reduce((a, b) => a + b.rows, 0),
          )} of the rows, and ${exact(d.tail_values)} more ${
            d.tail_values === 1 ? 'value holds' : 'values hold'
          } the rest.`
        : `Piled at the low end and falling away — ${share(top.rows)} of the rows are in the first bucket.`
    case 'mixed':
      return `Spread across ${occupiedCount(d)} of ${d.buckets.length} buckets, the fullest holding ${share(
        top.rows,
      )}.`
    case 'empty':
      return 'Nothing to read.'
  }
}

/** How many rows are missing from the picture, said rather than left to be
 *  noticed. A NULL is not a value and has no place on an axis of values, which
 *  is why it is excluded from the count and named here instead. */
export function asideFrom(d: Distribution): string | null {
  if (d.nulls === 0) return null
  const of = d.rows + d.nulls
  return `${exact(d.nulls)} of ${exact(of)} rows are null and are not on this axis.`
}

function listedRows(d: Distribution): number {
  return d.buckets.reduce((a, b) => a + b.rows, 0)
}

function occupiedCount(d: Distribution): number {
  return d.buckets.filter((b) => b.rows > 0).length
}

function peakLabels(d: Distribution): string[] {
  const total = d.rows
  return d.buckets
    .filter((b) => b.rows / total >= PEAK)
    .map((b) => (d.mode === 'bins' ? `around ${b.label}` : value(b.label, d)))
}

/** A bucket's label as it should read in a sentence. A binned bucket is a range
 *  and says so; a value is quoted, because `alpha` and alpha are different
 *  claims about what is in the column. */
function value(label: string, d: Distribution): string {
  return d.mode === 'bins' ? `around ${label}` : `\`${label}\``
}

function exact(n: number): string {
  return n.toLocaleString('en-GB')
}

/** The bars, scaled to the fullest bucket.
 *
 *  To the fullest rather than to the total: a column whose largest bucket is 3%
 *  of the table draws sixteen invisible bars against the total, and the shape —
 *  which is the entire point — disappears. The figure beside the chart says what
 *  the full height is, the same contract the sparklines keep. */
export function bars(d: Distribution): { label: string; rows: number; share: number }[] {
  const peak = d.buckets.reduce((m, b) => Math.max(m, b.rows), 0)
  return d.buckets.map((b) => ({
    label: b.label,
    rows: b.rows,
    share: peak > 0 ? b.rows / peak : 0,
  }))
}

File: src/clickhouse/distribution.rs
//! The shape of one column.
//!
//! The profile gives five numbers per column — distinct, null, min, max, mean —
//! and five numbers cannot say whether a column is evenly spread, piled at one
//! end, split into two clusters, or one value and a rounding error. That is the
//! difference between knowing a column's range and knowing its distribution, and
//! it is the last of the brief's analysis list that the product had no answer
//! for.
//!
//! This module **measures**: it counts rows into buckets and says nothing about
//! what the counts mean. Naming the shape and wording it is
//! `frontend/src/lib/distribution.ts`, where the thresholds are pure functions
//! with a test each — the split `review.rs` and `projection.rs` set.
//!
//! Asked for one column at a time. A table's worth would be one query per
//! column, and the profile already reads every column in a single pass for the
//! numbers that a single pass can produce; a histogram is not one of them,
//! because each column needs its own range before it can be binned.
//!
//! Four things were measured on real tables before this was written, and each
//! one decided part of what is counted.
//!
//! - **ClickHouse's own `histogram(n)` is the wrong tool here.** It answers in
//!   9 ms and returns *adaptive* bins: unequal widths and fractional heights —
//!   `(12, 208.08, 105473.75)` on a real column. Bars of equal width drawn from
//!   bins of unequal width misstate density, which is the one thing a histogram
//!   exists to show, and 105,473.75 is not a number of rows. Equal-width bins
//!   cost 15 ms over the same 482,212 rows and are exact.
//! - **Empty buckets vanish from a `GROUP BY`, and the axis lies.**
//!   `analytics.device_daily.events` binned into twelve came back as five rows —
//!   bins 0, 1, 6, 7 and 11. Drawn in order they read as five adjacent bars, a
//!   smooth ramp; the truth is 42,000 rows at the top of the range and 800
//!   scattered below it. `WITH FILL FROM 0 TO n STEP 1` emits the empties, the
//!   same way the drift reader gets its gaps.
//! - **A column with few distinct values is a tally, not a histogram.** That
//!   same column holds six distinct values across a range twelve wide, so seven
//!   of its twelve bins were empty *by construction* and the chart was reporting
//!   an artefact of the binning rather than a fact about the data. Below
//!   `TALLY_MAX` the values are counted as themselves.
//! - **A high-cardinality string has neither a range nor a tally.** It gets its
//!   most common values and an honest remainder: what is not listed is counted
//!   and said, never dropped.

use serde::{Deserialize, Serialize};

use super::{Client, QueryOptions};
use crate::error::{Error, Result};

/// At or below this many distinct values, the values are counted as themselves.
/// Above it there are more bars than a reader can tell apart anyway.
const TALLY_MAX: u64 = 32;

/// Buckets a binned column is cut into. Sixteen is enough to show a second peak
/// and few enough that each bar is still wide enough to see.
const BINS: u64 = 16;

/// Values listed for a column that is neither binnable nor small enough to
/// tally. The rest are counted into one remainder rather than dropped.
const TOP: u64 = 12;

/// How the counts were produced, which the reader above needs in order to know
/// what a bucket *is*.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Mode {
    /// One bucket per distinct value, all of them.
    Tally,
    /// Equal-width bins across the column's range, empties included.
    Bins,
    /// The most common values, with everything else counted into `tail_rows`.
    Top,
}

#[derive(Debug, Clone, Serialize)]
pub struct Bucket {
    /// What to print under the bar. For `bins` this is the lower edge, because a
    /// bin is named by where it starts.
    pub label: String,
    pub rows: u64,
    /// The edges, for a binned column only — so the reader can say how wide a
    /// bar is rather than leaving the axis to be guessed.
    pub from: Option<f64>,
    pub to: Option<f64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Distribution {
    pub available: bool,
    pub reason: Option<String>,
    pub column: String,
    pub r#type: String,
    pub mode: Mode,
    /// Rows the count was taken over, NULLs excluded — a NULL is not a value and
    /// has no place on an axis of values. How many there were is said separately.
    pub rows: u64,
    pub nulls: u64,
    pub distinct: u64,
    pub buckets: Vec<Bucket>,
    /// For `top`: the rows and the values that are not in `buckets`.
    pub tail_rows: u64,
    pub tail_values: u64,
}

pub async fn distribution(
    ch: &Client,
    database: &str,
    table: &str,
    column: &str,
) -> Result<Distribution> {
    #[derive(Deserialize)]
    struct Col {
        r#type: String,
    }
    let cols: Vec<Col> = ch
        .rows_with(
            "SELECT type AS type FROM system.columns \
             WHERE database = {db:String} AND table = {tbl:String} AND name = {col:String}",
            params(database, table, column),
        )
        .await?;
    let Some(Col { r#type }) = cols.into_iter().next() else {
        return Err(Error::BadRequest(format!(
            "`{column}` is not a column of `{database}`.`{table}`"
        )));
    };

    let q = quote(column);
    let from = format!("{}.{}", quote(database), quote(table));

    #[derive(Deserialize)]
    struct Facts {
        n: u64,
        nulls: u64,
        u: u64,
        lo: Option<f64>,
        hi: Option<f64>,
    }
    /* One pass for everything the choice of mode depends on. `min`/`max` are cast
    through Float64 only where the type allows it — asking a String for its range
    answers with a kilobyte of the first value, which is the trap `profile.rs`
    documents. */
    let range = if numeric(&r#type) || temporal(&r#type) {
        format!(
            "toFloat64OrNull(toString(min({q}))) AS lo, toFloat64OrNull(toString(max({q}))) AS hi"
        )
    } else {
        "NULL AS lo, NULL AS hi".to_string()
    };
    let facts: Vec<Facts> = match ch
        .rows_with(
            &format!(
                "SELECT toUInt64(count()) AS n, toUInt64(countIf({q} IS NULL)) AS nulls, \
                 toUInt64(uniqExact({q})) AS u, {range} FROM {from}"
            ),
            params(database, table, column),
        )
        .await
    {
        Ok(f) => f,
        Err(e) if refused(&e) => return Ok(empty(column, &r#type, Some(said(&e)))),
        Err(e) => return Err(e),
    };
    let Some(f) = facts.into_iter().next() else {
        return Ok(empty(column, &r#type, None));
    };
    if f.n == f.nulls {
        // Every row is NULL: there is no distribution, and drawing an empty axis
        // would suggest the question was answered.
        return Ok(Distribution {
            available: true,
            rows: 0,
            nulls: f.nulls,
            distinct: 0,
            ..empty(column, &r#type, None)
        });
    }

    let live = f.n - f.nulls;

    /* Small enough to be itself, or continuous enough to be binned, or neither.
    The order matters: a numeric column of six values is a tally, because six
    values across a range of twelve produce empty bins by construction and the
    chart would be reporting the binning rather than the data. */
    if f.u <= TALLY_MAX {
        let rows: Vec<serde_json::Value> = read(
            ch,
            &format!(
                "SELECT toString({q}) AS label, toUInt64(count()) AS n FROM {from} \
                 WHERE {q} IS NOT NULL GROUP BY {q} ORDER BY n DESC, label"
            ),
            database,
            table,
            column,
        )
        .await?;
        return Ok(Distribution {
            available: true,
            reason: None,
            column: column.to_string(),
            r#type,
            mode: Mode::Tally,
            rows: live,
            nulls: f.nulls,
            distinct: f.u,
            buckets: rows
                .iter()
                .map(|r| Bucket {
                    label: text(r, "label").unwrap_or_default(),
                    rows: num(r, "n"),
                    from: None,
                    to: None,
                })
                .collect(),
            tail_rows: 0,
            tail_values: 0,
        });
    }

    if let (true, Some(lo), Some(hi)) = (numeric(&r#type) || temporal(&r#type), f.lo, f.hi) {
        if hi > lo {
            let width = (hi - lo) / BINS as f64;
            let integral = whole(&r#type);
            /* `WITH FILL` for the same reason the drift reader needs it: a bin
            with no rows is omitted by `GROUP BY`, and a chart drawn from what
            came back closes the gap silently. The clamp on the last bin is
            because the maximum value lands one past the end by arithmetic. */
            let rows: Vec<serde_json::Value> = read(
                ch,
                &format!(
                    "SELECT bin, n FROM ( \
                       SELECT least({}, toUInt32(floor((toFloat64OrNull(toString({q})) - {lo}) / {width}))) AS bin, \
                              toUInt64(count()) AS n \
                       FROM {from} WHERE {q} IS NOT NULL GROUP BY bin \
                     ) ORDER BY bin WITH FILL FROM 0 TO {} STEP 1",
                    BINS - 1,
                    BINS
                ),
                database,
                table,
                column,
            )
            .await?;
            return Ok(Distribution {
                available: true,
                reason: None,
                column: column.to_string(),
                r#type,
                mode: Mode::Bins,
                rows: live,
                nulls: f.nulls,
                distinct: f.u,
                buckets: rows
                    .iter()
                    .map(|r| {
                        let i = num(r, "bin") as f64;
                        Bucket {
                            label: trim(lo + i * width, integral),
                            rows: num(r, "n"),
                            from: Some(lo + i * width),
                            to: Some(lo + (i + 1.0) * width),
                        }
                    })
                    .collect(),
                tail_rows: 0,
                tail_values: 0,
            });
        }
    }

    // Neither small nor continuous: the common values, and an honest remainder.
    let rows: Vec<serde_json::Value> = read(
        ch,
        &format!(
            "SELECT toString({q}) AS label, toUInt64(count()) AS n FROM {from} \
             WHERE {q} IS NOT NULL GROUP BY {q} ORDER BY n DESC, label LIMIT {TOP}"
        ),
        database,
        table,
        column,
    )
    .await?;
    let buckets: Vec<Bucket> = rows
        .iter()
        .map(|r| Bucket {
            label: text(r, "label").unwrap_or_default(),
            rows: num(r, "n"),
            from: None,
            to: None,
        })
        .collect();
    let listed: u64 = buckets.iter().map(|b| b.rows).sum();
    Ok(Distribution {
        available: true,
        reason: None,
        column: column.to_string(),
        r#type,
        mode: Mode::Top,
        rows: live,
        nulls: f.nulls,
        distinct: f.u,
        tail_rows: live.saturating_sub(listed),
        tail_values: f.u.saturating_sub(buckets.len() as u64),
        buckets,
    })
}

async fn read(
    ch: &Client,
    sql: &str,
    database: &str,
    table: &str,
    column: &str,
) -> Result<Vec<serde_json::Value>> {
    ch.rows_with(sql, params(database, table, column)).await
}

fn params(database: &str, table: &str, column: &str) -> QueryOptions {
    QueryOptions {
        params: vec![
            ("db".into(), database.to_string()),
            ("tbl".into(), table.to_string()),
            ("col".into(), column.to_string()),
        ],
        ..QueryOptions::internal()
    }
}

/// A bin's lower edge, as short as it can be without lying.
///
/// `208.08278830699174` is arithmetic, not a label. And an *integer* column has
/// no fractional edges to report at all: a `UInt32` holding 100 to 939 binned
/// into sixteen produced an edge of `886.56`, which is a number the column
/// cannot contain and a reader would have to unlearn.
fn trim(v: f64, whole: bool) -> String {
    if whole || (v == v.trunc() && v.abs() < 1e15) {
        format!("{}", v.round() as i64)
    } else {
        let s = format!("{v:.2}");
        s.trim_end_matches('0').trim_end_matches('.').to_string()
    }
}

/// Whether the column's own type has no fractions in it.
fn whole(declared: &str) -> bool {
    let t = strip(declared);
    t.starts_with("UInt") || t.starts_with("Int") || t.starts_with("Date")
}

fn strip(declared: &str) -> &str {
    declared
        .trim_start_matches("LowCardinality(")
        .trim_start_matches("Nullable(")
}

fn numeric(declared: &str) -> bool {
    let t = strip(declared);
    t.starts_with("UInt")
        || t.starts_with("Int")
        || t.starts_with("Float")
        || t.starts_with("Decimal")
}

fn temporal(declared: &str) -> bool {
    strip(declared).starts_with("Date")
}

fn quote(name: &str) -> String {
    format!("`{}`", name.replace('`', "``"))
}

fn num(row: &serde_json::Value, key: &str) -> u64 {
    row.get(key)
        .and_then(|v| {
            v.as_u64()
                .or_else(|| v.as_str().and_then(|s| s.parse().ok()))
        })
        .unwrap_or(0)
}

fn text(row: &serde_json::Value, key: &str) -> Option<String> {
    row.get(key).and_then(|v| v.as_str()).map(|s| s.to_string())
}

fn refused(e: &Error) -> bool {
    matches!(
        e,
        Error::ClickHouse { code: 497, .. }   // ACCESS_DENIED
            | Error::ClickHouse { code: 241, .. } // MEMORY_LIMIT_EXCEEDED
            | Error::ClickHouse { code: 159, .. } // TIMEOUT_EXCEEDED
    )
}

fn said(e: &Error) -> String {
    match e {
        Error::ClickHouse { code: 241, .. } => {
            "counting this column's values ran past the server's memory limit".to_string()
        }
        Error::ClickHouse { code: 159, .. } => "the scan ran past the query timeout".to_string(),
        other => other.to_string().lines().next().unwrap_or("").to_string(),
    }
}

fn empty(column: &str, declared: &str, reason: Option<String>) -> Distribution {
    Distribution {
        available: reason.is_none(),
        reason,
        column: column.to_string(),
        r#type: declared.to_string(),
        mode: Mode::Tally,
        rows: 0,
        nulls: 0,
        distinct: 0,
        buckets: Vec::new(),
        tail_rows: 0,
        tail_values: 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_type_is_read_through_its_wrappers() {
        assert!(numeric("Nullable(UInt32)"));
        assert!(numeric("LowCardinality(Int8)"));
        assert!(!numeric("LowCardinality(String)"));
        assert!(temporal("Nullable(DateTime64(3))"));
        assert!(!temporal("String"));
    }

    #[test]
    fn a_bin_edge_is_a_label_rather_than_arithmetic() {
        // `208.08278830699174` is what the division produces and nothing anyone
        // wants under a bar.
        assert_eq!(trim(208.082_788_306_991_74, false), "208.08");
        assert_eq!(trim(12.0, false), "12");
        assert_eq!(trim(-3.5, false), "-3.5");
        assert_eq!(trim(1.10, false), "1.1");
        // An integer column has no fractional edges to report at all.
        assert_eq!(trim(886.56, true), "887");
        assert!(whole("UInt32") && whole("Nullable(Date)") && !whole("Float64"));
    }

    #[test]
    fn an_identifier_survives_a_backtick() {
        assert_eq!(quote("a`b"), "`a``b`");
    }
}
